### DISCUSSION TITLE
Calling functions on a rust object

### DISCUSSION THREAD

**Discussion by eyyyyyyy3** [tauri-apps/tauri#8365] [Q&A] [answered]
Is it possible to have a function that can be passed to the frontend and that can be called with invoke that works on a instance of a struct in rust?

Let's take the following code as a example:
>
    struct Connection {
        is_connected: bool,
        handle: Option<Client>,
    }

    impl Connection {
        #[tauri::command]
        pub fn establish_connection(address: &str, port: u16) -> Self {
            let ret = match Client::connect(&format!("host={} port={}", address, port,), NoTls) {
                Ok(x) => Connection {
                    is_connected: true,
                    handle: Some(x),
                },
                Err(_) => Connection {
                    is_connected: false,
                    handle: None,
                },
            };
            ret
        }
    }

Lets ignore the fact, that this isn't compiling (because the attribute causes some errors inside of a impl block "error: `use` import is not supported in `trait`s or `impl`s" ). How and where would the instance be stored that is created by establish_connection and how could I modify that special instance? To be more precise how would I do a `fn example_that_changes_internal_values(&mut self) {}` inside of the frontend.

Why do I want to do this? It's simple. I want to be able to create a instance of struct and have all its fields at hand to be able to manipulate them or work with them without exposing too much to the frontend because it isn't of any real importance to the frontend.

**Comment by FabianLars** [comment]
So this doesn't work _directly_. The closest thing to that is to store the Connection instance in `tauri::State` and have a "normal" command that uses that state to call the instance method.
https://docs.rs/tauri/latest/tauri/trait.Manager.html#method.manage

**Reply by eyyyyyyy3** [reply]
How would a function call with a state as a parameter of that function look like in the front end?

**Reply by FabianLars** [reply, marked as answer]
The same as without state. Tauri will inject the state for you. So for example both of these commands are the same on the javascript side
```rs
fn my_cmd() {}
fn my_cmd(state: State<SomeType>) {}
```
-> `invoke("my_cmd")`
